### PR TITLE
ddl: check column data when execute alter table/column charset

### LINKED ISSUE #34008: Unsupported modify charset from latin1 to utf8mb4
## Bug Report

<p style="margin: 10px 0px 0px; padding: 0px; color: rgb(23, 43, 77); font-family: -apple-system, &quot;system-ui&quot;, &quot;Segoe UI&quot;, Roboto, Oxygen, Ubuntu, &quot;Fira Sans&quot;, &quot;Droid Sans&quot;, &quot;Helvetica Neue&quot;, sans-serif; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">This can be easily simulated<br>MySQL<span> </span><span class="error">[test]</span>&gt; create table t2(a varchar(20) )CHARSET=latin1;<br>MySQL<span> </span><span class="error">[test]</span>&gt; insert into t2 values('&lt;some value&gt;');<br>MySQL<span> </span><span class="error">[test]</span>&gt; alter table t2 DEFAULT CHARACTER SET utf8mb4;<br>ERROR 8200 (HY000): Unsupported modify charset from latin1 to utf8mb4</p><p style="margin: 10px 0px 0px; padding: 0px; color: rgb(23, 43, 77); font-family: -apple-system, &quot;system-ui&quot;, &quot;Segoe UI&quot;, Roboto, Oxygen, Ubuntu, &quot;Fira Sans&quot;, &quot;Droid Sans&quot;, &quot;Helvetica Neue&quot;, sans-serif; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">If we create another table with charset=utf8mb4 and insert from t2. There is no error<br>MySQL<span> </span><span class="error">[test]</span>&gt; create table t3 (a varchar(20)) charset=utf8mb4;<br>MySQL<span> </span><span class="error">[test]</span>&gt; insert into t3 select * from t2;<br>Query OK, 1 row affected (0.01 sec)<br>MySQL<span> </span><span class="error">[test]</span>&gt; select * from t3;<br><ins>------</ins></p><div class="table-wrap" style="margin: 0px; padding: 0px; color: rgb(23, 43, 77); font-family: -apple-system, &quot;system-ui&quot;, &quot;Segoe UI&quot;, Roboto, Oxygen, Ubuntu, &quot;Fira Sans&quot;, &quot;Droid Sans&quot;, &quot;Helvetica Neue&quot;, sans-serif; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">

a
--

</div><p style="margin: 10px 0px 0px; padding: 0px; color: rgb(23, 43, 77); font-family: -apple-system, &quot;system-ui&quot;, &quot;Segoe UI&quot;, Roboto, Oxygen, Ubuntu, &quot;Fira Sans&quot;, &quot;Droid Sans&quot;, &quot;Helvetica Neue&quot;, sans-serif; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><ins>------</ins><br>It is not clear if the above insert would convert latin char to utf8 correctly.</p><p style="margin: 10px 0px 0px; padding: 0px; color: rgb(23, 43, 77); font-family: -apple-system, &quot;system-ui&quot;, &quot;Segoe UI&quot;, Roboto, Oxygen, Ubuntu, &quot;Fira Sans&quot;, &quot;Droid Sans&quot;, &quot;Helvetica Neue&quot;, sans-serif; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">Customer is aware of the latin support issue as documented and in<span> </span><a href="https://github.com/pingcap/tidb/issues/18955" class="external-link" target="_blank" rel="nofollow noopener" style="color: rgb(0, 82, 204); text-decoration: var(--aui-link-decoration); cursor: pointer;">https://github.com/pingcap/tidb/issues/18955</a></p>

### 1. Minimal reproduce step (Required)
MySQL [test]> create table t2(a varchar(20) )CHARSET=latin1;
MySQL [test]> insert into t2 values('<some value>');
MySQL [test]> alter table t2 DEFAULT CHARACTER SET utf8mb4;

### 2. What did you expect to see? (Required)
Query OK
### 3. What did you see instead (Required)
ERROR 8200 (HY000): Unsupported modify charset from latin1 to utf8mb4
### 4. What is your TiDB version? (Required)
v5.2.2

### LINKED ISSUE #34562: Support check invalid utf8 characters during `Alter table ... charset` stmts

### PR DESCRIPTION
**Pull request by Defined2014**
### What problem does this PR solve?

Issue Number: close #34562

Problem Summary:

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
An error will be reported when executing a alter charset statement with invalid data.
```

### CODE REVIEW

File: parser/charset/encoding_latin1.go [hunk 1]
```diff
@@ -14,6 +14,8 @@
 package charset
 
 import (
+	"bytes"
```
Comment (hawkingrei) [inline parser/charset/encoding_latin1.go:17]: Please use ```goimports``` to format.

File: ddl/column.go [hunk 2]
```diff
@@ -1556,6 +1557,53 @@ func updateChangingObjState(changingCol *model.ColumnInfo, changingIdxs []*model
...
+	sctx, err := w.sessPool.get()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	sctx.GetSessionVars().StmtCtx.TruncateAsWarning = false
+	sctx.GetSessionVars().StmtCtx.IgnoreTruncate = false
+	defer w.sessPool.put(sctx)
+
+	var buf strings.Builder
+	buf.WriteString("select * from %n.%n where ")
```
Comment (tangenta) [inline ddl/column.go:1577]:
```suggestion
	buf.WriteString("select 1 from %n.%n where ")
```

File: ddl/column.go [hunk 3]
```diff
@@ -1556,6 +1557,53 @@ func updateChangingObjState(changingCol *model.ColumnInfo, changingIdxs []*model
 	}
 }
 
+func needCheckLatin1Convert(oldCol, newCol *model.ColumnInfo) bool {
+	if oldCol.GetCharset() != newCol.GetCharset() {
+		return oldCol.GetCharset() == charset.CharsetLatin1
+	}
+	return false
```
Comment (tangenta) [inline ddl/column.go:1564]:
```suggestion
	return oldCol.GetCharset() == charset.CharsetLatin1 && newCol.GetCharset() != charset.CharsetLatin1
```

File: parser/go.mod [hunk 4]
```diff
@@ -7,10 +7,12 @@ require (
 	github.com/go-sql-driver/mysql v1.6.0
 	github.com/pingcap/errors v0.11.5-0.20210425183316-da1aaba5fb63
 	github.com/pingcap/log v0.0.0-20210625125904-98ed8e2eb1c7
+	github.com/pkg/errors v0.9.1 // indirect
```
Comment (tangenta) [inline parser/go.mod:10]: Please run `go mod tidy` in the parser directory.

File: ddl/table.go [hunk 5]
```diff
@@ -1040,11 +1041,32 @@ func onModifyTableCharsetAndCollate(d *ddlCtx, t *meta.Meta, job *model.Job) (ve
...
+			if err != nil {
+				if job.ReorgMeta.SQLMode.HasStrictMode() {
+					job.State = model.JobStateCancelled
+					return ver, err
+				} else {
+					warn := errors.Cause(err).(*terror.Error)
+					job.ReorgMeta.Warnings[warn.ID()] = warn
+					job.ReorgMeta.WarningsCount[warn.ID()] = 1
+				}
+			}
```
Comment (zimulala) [inline ddl/table.go:1068]: We can extract the above code to deal with this problem(add a jobState argument), then we can use this function in https://github.com/pingcap/tidb/pull/34534/files#diff-42e0180cdc8c14def674a9e45d908d6700f6a8ccba0f9ff4b14b27c0daa49859R1640-R1650 too.

File: ddl/column.go (hunk 2 above)
Comment (Defined2014) [reply ddl/column.go]: We need the result for generate error msg.

File: ddl/table.go (hunk 5 above)
Comment (Defined2014) [reply ddl/table.go]: Done

File: ddl/db_integration_test.go [hunk 6]
```diff
@@ -909,6 +909,50 @@ func TestChangingTableCharset(t *testing.T) {
...
+	tk.MustExec("drop table if exists t")
+	tk.MustExec("create table t(a char(20) charset latin1, b char(20), c char(20) charset latin1)")
+	tk.MustExec("insert into t values (0x70, 0x61, 0x90)")
+	tk.MustGetErrMsg("alter table t convert to charset utf8mb4", "[table:1366]Incorrect string value '\\x90' for column 'c'")
+
+	tk.MustExec("set sql_mode=''")
+	tk.MustExec("drop table if exists t")
+	tk.MustExec("create table t (a varchar(20)) charset = latin1")
+	tk.MustExec("insert into t values (0x90)")
+	tk.MustExec("alter table t convert to charset utf8")
```
Comment (zimulala) [inline ddl/db_integration_test.go:941]: If I execute this statement locally, I get an error:
```
tidb>   alter table t convert to charset utf8;
ERROR 8200 (HY000): Unsupported modify charset from utf8mb4 to utf8
```
`alter table t convert to charset utf8mb4;` is OK.

Comment (Defined2014) [reply ddl/db_integration_test.go:941]: If you have a row which charset is utf8mb4, the `alter` stmt will report an error.

File: ddl/db_integration_test.go [hunk 7]
```diff
@@ -909,6 +909,50 @@ func TestChangingTableCharset(t *testing.T) {
...
+	tk.MustQuery("show warnings;").Check(testkit.Rows("Warning 1366 Incorrect string value '\\x90' for column 'a'"))
+	tk.MustQuery("select hex(a) from t").Check(testkit.Rows("90"))
+	tk.MustQuery("show create table t;").Check(testkit.Rows(
+		"t CREATE TABLE `t` (\n  `a` varchar(20) DEFAULT NULL\n) ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_bin"))
```
[the remark this answers is not in the document]
Comment (Defined2014) [reply ddl/db_integration_test.go:945]: Done

File: ddl/db_integration_test.go (hunk 6 above)
Comment (zimulala) [reply ddl/db_integration_test.go:946]: Got it.

File: ddl/table.go [hunk 8]
```diff
@@ -1040,11 +1040,26 @@ func onModifyTableCharsetAndCollate(d *ddlCtx, t *meta.Meta, job *model.Job) (ve
...
+			if field_types.HasCharset(&col.FieldType) && needCheckLatin1Convert(col, newCol) {
+				oldCols = append(oldCols, col)
+				newCols = append(newCols, newCol)
+			}
+		}
+		if len(oldCols) != 0 {
+			err := w.checkLatin1Convert(job, dbInfo.Name, tblInfo.Name, oldCols, newCols)
+			if err != nil {
+				job.State = model.JobStateCancelled
+				return ver, err
```
Comment (zimulala) [inline ddl/table.go:1060]:
```suggestion
				return ver, errors.Trace(err)
```

**Review by zimulala — APPROVED** [review]
LGTM

**Comment by Defined2014** [comment]
This is not a good implementation, close it first.

**Pull request closed without merging**